File: backend/app/schemas/resume.py
"""Pydantic schemas for the Resume Intelligence API (spec sections 6, 27)."""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.resume import ResumeStatus, SectionType


class ResumeRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    original_filename: str
    file_type: str
    file_size_bytes: int
    status: ResumeStatus
    failure_reason: str | None = None
    created_at: datetime
    updated_at: datetime


class ResumeSectionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    section_type: SectionType
    order_index: int
    content: str


class ScoreExplanation(BaseModel):
    """Every sub-score's explanation is a free-form dict because the checks
    behind ATS/Technical/Skill scores differ in shape from the AI-grounded
    reasoning behind Experience/Project scores — see
    app/services/resume/scoring.py and ai_review.py for what each contains."""

    model_config = ConfigDict(extra="allow")


class ResumeAnalysisRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    resume_id: uuid.UUID
    overall_score: int
    ats_score: int
    technical_score: int
    experience_score: int
    project_score: int
    skill_score: int
    score_explanations: dict
    detected_skills: list
    missing_keywords: list
    ats_issues: list
    strengths: list
    weaknesses: list
    analysis_mode: str
    created_at: datetime


class ResumeDetailRead(ResumeRead):
    sections: list[ResumeSectionRead] = []

File: frontend/src/types/resume.ts
export type ResumeStatus = "pending" | "processing" | "completed" | "failed";

export interface Resume {
  id: string;
  original_filename: string;
  file_type: "pdf" | "docx";
  file_size_bytes: number;
  status: ResumeStatus;
  failure_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface ResumeSection {
  section_type: string;
  order_index: number;
  content: string;
}

export interface ResumeDetail extends Resume {
  sections: ResumeSection[];
}

export interface DetectedSkill {
  skill: string;
  category: string;
  tier: number;
  evidence: string;
  line_number: number;
}

export interface ResumeAnalysis {
  id: string;
  resume_id: string;
  overall_score: number;
  ats_score: number;
  technical_score: number;
  experience_score: number;
  project_score: number;
  skill_score: number;
  score_explanations: Record<string, unknown>;
  detected_skills: DetectedSkill[];
  missing_keywords: string[];
  ats_issues: string[];
  strengths: string[];
  weaknesses: string[];
  // "deterministic" = rule-based only (no AI provider configured, or the
  // AI call failed). "ai_assisted" = experience/project scores and
  // narrative feedback came from the configured LLM.
  analysis_mode: "deterministic" | "ai_assisted";
  created_at: string;
}
